File: cpp/n14_longestPrefix.cpp
// https://leetcode-cn.com/problems/longest-common-prefix/
#include <string>
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    string longestCommonPrefix(vector<string>& strs) {
        auto iter = strs.begin();
        if (iter == strs.end()) return "";
        string s = *iter;
        while(++iter != strs.end()){          
            string t = *iter;
            while(t.find(s) != 0) s.pop_back();
        }
        return s;
    }
};

int main(){
    Solution solu;
    vector<string> strs = {"flower","flow","flight"};
    cout << solu.longestCommonPrefix(strs) << endl;
}

File: python/n14_longestPrefix.py
# https://leetcode-cn.com/problems/longest-common-prefix/
class Solution:
    def longestCommonPrefix(self, strs) -> str:     
        s = ""
        for i in zip(*strs):
            if len(set(i)) == 1:
                s += i[0]
            else:
                break           
        return s
    
    def longestCommonPrefix_find(self, strs) -> str:
        if not strs:
            return ""
        res = strs[0]
        for i in range(len(strs)):
            while strs[i].find(res) != 0:
                res = res[:-1]
        return res

if __name__ == "__main__":
    solu = Solution()
    test_lis = ["flower","flow","flight"]
    print(list(zip(*test_lis)))
    print(solu.longestCommonPrefix(test_lis))
    print(solu.longestCommonPrefix_find(test_lis))